### PR TITLE
First pass adding pure PWA

### PR DESCRIPTION
**Pull request by gravityrail**
Adds some basic PWA features:
- cache all non-WP-Admin responses retrieved via GET
- add manifest for pseudo-app-install and splash screen support
- detect theme colors
- Official AMP plugin integration

### CODE REVIEW

File: modules/pwa/assets/css/show-network-status.css [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#jetpack-network-status {
+    display: none;
+}
+
+body.jetpack__offline #jetpack-network-status {
+    display: block;
+}
```
Comment (ebinnion) [inline modules/pwa/assets/css/show-network-status.css]: Minor drive-by nitpick. This, and the other files below, don't end with an empty newline.

**Comment by jeherve** [comment]
Do you think you could add some screenshots and some testing instructions so we all know what to look for?

Thanks!

File: modules/pwa/pwa.php [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+add_action( 'jetpack_activate_module_pwa', 'pwa_activate' );
+add_action( 'jetpack_deactivate_module_pwa', 'pwa_deactivate' );
+
+// register WP_Query hooks for manifest and service worker
+add_filter( 'query_vars', 'pwa_register_query_vars' );
+
+// manifest
+add_action( 'wp_head', 'pwa_render_manifest_link' );
+// add_action( 'admin_head', 'pwa_render_manifest_link' ); // Don't load for wp-admin, for now
+add_action( 'amp_post_template_head', 'pwa_render_manifest_link' ); // AMP
```
Comment (jeherve) [inline modules/pwa/pwa.php]: How does this play with other AMP plugins like our own? Shouldn't we not output anything when that plugin is active?

File: modules/pwa/pwa.php [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+// service worker
+add_action( 'template_redirect', 'pwa_force_https', 1 );
```
Comment (jeherve) [inline modules/pwa/pwa.php]: How does this work on sites that do not have an active SSL certificate?

Comment (gravityrail) [reply modules/pwa/pwa.php]: It doesn't work at all on those sites.

File: modules/pwa/pwa.php (hunk 2 above)
Comment (gravityrail) [reply modules/pwa/pwa.php]: `amp_post_template_head` is an action fired by our AMP plugin. AMP needs a specific manifest format that is not regular HTML. This feature is designed to work with the official AMP plugin, not replace it.

**Comment by gravityrail** [comment]
> Do you think you could add some screenshots and some testing instructions so we all know what to look for?

I will soon! This is still a work in progress.

File: modules/pwa/assets/css/show-network-status.css (hunk 1 above)
Comment (gravityrail) [reply modules/pwa/assets/css/show-network-status.css]: Fixed, thanks. Will push up this change soon.

File: _inc/lib/class.core-rest-api-endpoints.php [hunk 4]
```diff
@@ -1394,6 +1420,51 @@ public static function get_updateable_data_list( $selector = '' ) {
...
+			),
+			'pwa_show_network_status' => array(
+				'description'       => esc_html__( 'Display notice on page when the browser is offline', 'jetpack' ),
+				'type'              => 'boolean',
+				'default'           => 0,
+				'validate_callback' => __CLASS__ . '::validate_boolean',
+				'jp_group'          => 'pwa',
+			),
+			'pwa_lazy_images' => array(
+				'description'       => esc_html__( 'Display notice on page when the browser is offline', 'jetpack' ),
```
Comment (beaulebens) [inline _inc/lib/class.core-rest-api-endpoints.php]: Dupe of the above label :)

File: modules/pwa/assets/js/service-worker.js [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+
+function fetchAndCache( request, event ) {
+    // open cache
+    return caches.open(CACHE).then( function( cache ) {
+		// find in cache
+
+		return cache.match( request )
+			.then( function( response ) {
+				// only do more work if we don't already have a response
+				if ( response ) {
```
Comment (igrigorik) [inline modules/pwa/assets/js/service-worker.js]: What's the invalidation process? If visitor caches content and site owner updates the resource.. Unlike the "regular" HTTP cache, resources in SW cache are not automatically checked for freshness and have to be manually (re+in)validated by our code here.

File: modules/perf/class.jetpack-perf-optimize-assets.php [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+		} elseif ( strpos( $registration->src, $site_url ) === 0 ) {
+			$registration->extra['jetpack-inline'] = true;
+			$raw_path = substr( $registration->src, strlen( $site_url ) );
+			$registration->extra['jetpack-inline-file'] = untrailingslashit( ABSPATH ) . $raw_path;
+		}
+
+		return isset( $registration->extra['jetpack-inline'] ) && $registration->extra['jetpack-inline'];
+	}
+
+	private function should_remove_style( $handle ) {
```
Comment (igrigorik) [inline modules/perf/class.jetpack-perf-optimize-assets.php]: For locally defined (inlined fonts), a better solution might be to inject a font-display [1] swap? For external resources, like Typekit and Google Fonts, that won't work today, but hopefully soon, once implementations for [2] land in Chrome and other browsers.
Alternatively, have you considered non-blocking font load / loadCSS [3] variant instead? That introduces FOUT, but I still think it's better than omitting them entirely? 

---

[1] https://www.w3.org/TR/css-fonts-4/#font-display-desc
[2] https://www.w3.org/TR/css-fonts-4/#font-display-font-feature-values
[3] https://github.com/filamentgroup/loadCSS#recommended-loadcss-usage

File: modules/pwa/assets/js/service-worker.js (hunk 5 above)
Comment (gravityrail) [reply modules/pwa/assets/js/service-worker.js]: I added stale-while-revalidate in the latest couple of commits. I'm not actually taking over the browser and replacing content yet, that's next...

**Comment by caraya** [comment]
Localhost is a secure origin so developers can work without a server or with a local server without a domain name. Currently, Jetpack disables all PWA features when working in development mode.

File: modules/pwa/class.jetpack-pwa-service-worker.php [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+			// fake localize - service worker is not loaded in page context, so regular localize doesn't work
+			$pwa = Jetpack_PWA::instance();
+			$pwa_vars = array(
+				'admin_url' => admin_url(),
+				'site_url' => site_url(),
+				'site_icon' => $pwa->site_icon_url( 48 ),
+				'sw_config_url' => $this->get_service_worker_config_url(),
+				'images_url' => plugins_url( 'assets/images/', __FILE__ )
+			);
+			echo preg_replace( '/pwa_vars_json/', json_encode( $pwa_vars ), file_get_contents( plugin_dir_path( __FILE__ ) . 'assets/js/service-worker.js' ) );
```
Comment (westonruter) [inline modules/pwa/class.jetpack-pwa-service-worker.php]: Maybe this has been discussed already, but I suggest adding `do_action( 'jetpack_print_service_worker_scripts' )` here so that plugins can add their own scripts, since Jetpack is reserving the `/service-worker.js` endpoint. This action would be analogous to the existing `wp_print_scripts` action.

Otherwise, the only alternative would be for a plugin to add a `template_redirect` action at priority 1 to print something before this `render_service_worker_js` method gets called.

Comment (westonruter) [reply modules/pwa/class.jetpack-pwa-service-worker.php]: Also, instead of doing a `preg_replace()` here which seems somewhat brittle, why not just do:

```php
printf( 'var pwa_vars_json = %s;', wp_json_encode( $pwa_vars ) );
echo file_get_contents( plugin_dir_path( __FILE__ ) . 'assets/js/service-worker.js' );
```

The `pwa_vars_json` variable is already being listed as a global.

File: modules/perf/class.jetpack-perf-optimize-assets.php [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+
+				$path = implode( DIRECTORY_SEPARATOR, $prefix ) . $path;
+			}
+
+			$dependency->extra['jetpack-inline-file'] = $path;
+		}
+
+		$should_inline = isset( $dependency->extra['jetpack-inline'] ) && $dependency->extra['jetpack-inline'];
+
+		$will_inline = apply_filters( $filter, $should_inline, $dependency->handle, $dependency->src ) && file_exists( $dependency->extra['jetpack-inline-file'] );
```
Comment (jeherve) [inline modules/perf/class.jetpack-perf-optimize-assets.php]: `apply_filters( $filter`

That kind of syntax is going to be a problem for our parser to handle. Maybe we could do something like what Core does with dynamic values inside a filter? Here is an example:
https://developer.wordpress.org/reference/hooks/_wp_post_revision_field_field/

File: modules/perf/class.jetpack-perf-optimize-assets.php [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+
+		return $will_inline;
+	}
+
+	private function should_remove_asset( $filter, $dependency ) {
+		return apply_filters( $filter, false, $dependency->handle, $dependency->src );
```
Comment (jeherve) [inline modules/perf/class.jetpack-perf-optimize-assets.php]: Same as above.

**Comment by caraya** [comment]
Is there any particular reason for using protocol relative and not HTTPS URLs when connecting to CDNs? There is no reason not to make the CDN URL an HTTPS one

File: modules/pwa/class.jetpack-pwa-service-worker.php [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+	}
+
+	public function render_amp_serviceworker_element() {
+		?>
+			<amp-install-serviceworker src="<?php echo $this->get_service_worker_url() ?>" layout="nodisplay"></amp-install-serviceworker>
+		<?php
+	}
+
+	private function get_service_worker_url() {
+		return add_query_arg( PWA_SW_QUERY_VAR, '1', trailingslashit( site_url() ) . 'index.php' );
```
Comment (westonruter) [inline modules/pwa/class.jetpack-pwa-service-worker.php]: If pretty permalinks are enabled, shouldn't this return `home_url( '/service-worker.js' )` as the rewrite rule was added in the `register_rewrite_rule` method?

**Comment by dereksmart** [comment]
@gravityrail can this be closed?
